### DISCUSSION TITLE
PR reviewing process proposal

### DISCUSSION THREAD

**Discussion by adizere** [informalsystems/hermes#2434] [General]
## Context

This is a preliminary discussion towards fixing https://github.com/informalsystems/ibc-rs/issues/2402.

I offer a perspective here that might be missing some parts. If we manage to agree on those missing parts and how to fix them, we can formalize everything in the CONTRIBUTING.md doc & close #2402.

## Table of contents:

1. A naive description of how PR reviewing is done using __reviewers__ only
2. Adding __assignees__ to the picture
3. Takeaways

### 1. A naive description of how PR reviewing is done, using only __reviewers__

In the figure below I describe a naive view of the PR reviewing process.

![Screenshot 2022-07-19 at 12 44 35](https://user-images.githubusercontent.com/1757002/179732329-6eccfe1e-f122-4787-9541-720c54e9ff13.png)

The main problem with this description is that the reviewers (which are tagged automatically) might:
- be away
- have no knowledge of the changed code
- have no context on how to reproduce the original issue
- have other priorities.

When we have stale issue, the main reasons for that are: 
a. we forgot to ask someone for explicit feedback, i.e., the PR needs further testing & reviewing, but it wasn't clear who should do it because there are multiple people tagged as reviewers
b. decisions needs to be made before the PR can progress, but neither the PR author nor reviewers feel like they can take those decisions.

Essentially, __reviewers are a best-effort__ activity. It works in 90% of PRs, but not always. Reviewers are only in charge with correctness of the PR, not with taking decisions out of the scope of correctness, nor are they responsible for timely merging, for opening subsequent issues or anything else.

### 2. Adding __assignees__ to the picture

Assignees are different from reviewers: a project maintainer manually tags the assignee, for instance. So an assigned person is not providing their best-effort, but rather a prioritized activity. The decision to assign someone will consequently shift people priorities, so it must take into account the context of the whole project's roadmap.

![Screenshot 2022-07-19 at 13 30 12](https://user-images.githubusercontent.com/1757002/179739958-1bb2b238-c332-4d71-96cf-b9c1e7c0d268.png)

I propose that an assignee acts as a person responsible for all things needed to close the PR in a timely manner. This is a big responsibility that is a super-set of the responsibilities of _both_ reviewer and PR author. Whenever a reviewer is unavaiable, the assignee will act as reviewer. If the PR author is away of has other priorities, the assignee will fix problems & apply review suggestions & handle the merging.

The assignee is a "safety net," and it involves a lot of responsibility. The project maintainers will tag assignees when necessary.

### Takeaways

**How many approvals are necessary to merge a PR?**
- Default, 1.
- In special cases, we signal specifically in the PR (one example is https://github.com/informalsystems/ibc-rs/pull/2275)

**What is the responsibility of reviewers?**
- Correctness of the PR
    - Which include testing (manually if necessary) and reviewing the code.
- They can ask for fixes, including new test cases.
- They can optionally do fixes (instead of the PR author).
- They can optionally merge the PR (instead of the PR author).

**What are the responsibilities of the PR author?**
- Fix review suggestions and provide rationale for the technical choices they made in their PR.
- Merge the PR once they have at least 1 approving review.
- No need to tag assignees nor reviewers, though they are encouraged to do that.

**What are the responsibilities of the assignee?**
- Testing, reviewing, "cat herding" reviewers and the PR author.
- Opening subsequent issue to capture follow-up work.
- Take any decisions needed to timely merge the PR, including coordination with other teams, for example.
- Fix problems in the PR (instead of the PR author), if necessary.
- Merging the PR.

**Comment by seanchen1991** [comment]
Transcribing the takeaways from today's weekly meeting here: 
- AZ: Important to tell explicitly the assigned person. Otherwise it gets lost in emails, buried.
    - PR owner should be first “line of offense” to tag reviewer or assigned.
- HS: Consider Slack integration with GitHub. Can get lost, must be personalized per person (DM).
- LJ: In any case, it should be documented. Not obvious it’s ok to DM people or assign them, explicit documentation.
    - AZ: Part of onboarding.
- SEC: Rust project. Bot randomly tags someone. Either “by luck” they are good, or they tag someone themselves.
- AZ proposal: rid of CODEOWNER. Any PR author tries to find someone, otherwise they tag someone. Project maintainers will be responsible.
- SA: It’s important to clarify the “responsibilities” (championing, get it merged, fix & solutions) of whoever is tagged reviewers.
    - Document this process.

**Comment by adizere** [comment]
Resolved in https://github.com/informalsystems/ibc-rs/pull/2438.

In particular:

https://github.com/informalsystems/ibc-rs/blob/f534cbd506f10fcc379bb161ecbffc88dadd3968/CONTRIBUTING.md?plain=1#L178-L215
